### ISSUE TITLE
[MNG-6972] Allow access to org.apache.maven.graph

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8062]
**[Michael Kroll](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickroll)** opened **[MNG-6972](https://issues.apache.org/jira/browse/MNG-6972?redirect=false)** and commented

Hi

maven doesn't export org.apache.maven.graph package in maven-core/src/main/resources/META-INF/maven/extension.xml so the GraphBuilder is not usable in extensions.

```java
// leads to java.lang.NoClassDefFoundError: Lorg/apache/maven/graph/GraphBuilder;
@Requirement( hint = GraphBuilder.HINT )
private GraphBuilder graphBuilder;
```

Background: if one build extension adds dependencies and another build extension uses `session.getProjectDependencyGraph()`, then the dependency graph is out of date. This is because the graph is only rebuilt after executing _all_ extensions. One solution to this would be to update the `MavenSession` and setting the new dependency graph in the first extension, but for this we need access to the `GraphBuilder`.

---

**Affects:** 3.6.3

**Issue Links:**
- [MNG-6974](https://issues.apache.org/jira/browse/MNG-6974) IT for MNG-5175 (Wagon HTTP) fails with JDK 14
 (_**"is blocked by"**_)

**Remote Links:**
- [GitHub Pull Request #69
](https://github.com/apache/maven-integration-testing/pull/69)
- [GitHub Pull Request #94
](https://github.com/apache/maven-integration-testing/pull/94)
- [GitHub Pull Request #368
](https://github.com/apache/maven/pull/368)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.9.0](https://github.com/apache/maven/milestone/2?closed=1), [3.8.8](https://github.com/apache/maven/milestone/109?closed=1)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

PRs with an IT are welcome.

**Comment by jira-importer** [comment]
**[Michael Kroll](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickroll)** commented

Added PR (one-liner).
Added IT.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [608744443012842b80a60fb257e291df922371ca](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=608744443012842b80a60fb257e291df922371ca) and [5cfcdbcda0ff9081dbb51bb162c2fd70fc177a9a](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=5cfcdbcda0ff9081dbb51bb162c2fd70fc177a9a).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet), [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), worthwhile for 3.9.0?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [a55300a588d3694966fd797323304e1ed0bc6de1](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=a55300a588d3694966fd797323304e1ed0bc6de1) for `maven-3.9.x` branch.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [43dffb925f23d46d38bda3a7f6949679a9e0da3b](https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=43dffb925f23d46d38bda3a7f6949679a9e0da3b) for `maven-3.8.x` branch.
